### ISSUE TITLE
When error during json writing, a registered error handler can't write proper error response - `ctx.json()` write error

### ISSUE THREAD

**Issue by kevin70** [avaje/avaje-http#112]
`ctx.json` Unable to output correct error information when an exception occurs during JSON serialization.
I created a project that contains sample code. [https://github.com/kevin70/avaje-jex-http-issues](https://github.com/kevin70/avaje-jex-http-issues)

Accessing the `/hello` interface will result in an illegal JSON data.
```
{
  "name": "Hello",
  "person": {{
  "error": "An error response"
}
```

Whether the expectation should be a correct error description?
like this?
```
{
  "error": "An error response"
}
```

**Comment by SentryMan** [comment]
If you define a custom exception handler for your server, the serialization exception will be caught and you can define your own error message. 

@rbygrave want to take a look at this? This also happens when we use Jsonb serialization to write to OutputStream. When the exception gets thrown during serialization we get an incomplete JSON.

```
{
    "name": "Hello",
    "person": {
        {
            "title": "Server Error",
            "status": 500,
            "type": "https://javalin.io/documentation#internalservererrorresponse",
            "details": {}
        }
```
Not sure how/if you want to tackle this.

**Comment by rbygrave** [comment]
So looking at Nima with the default Json handler it actually does write to the OutputStream BUT it may or may not flush. Nima actually detects if the output stream buffer has been written to or not and can give an error like:

```
Dec 12, 2022 4:41:55 PM org.eclipse.yasson.internal.Marshaller marshall
SEVERE: Unable to serialize property 'comments' from org.example.Person
Dec 12, 2022 4:41:55 PM org.eclipse.yasson.internal.Marshaller marshall
SEVERE: Generating incomplete JSON
Dec 12, 2022 4:42:33 PM io.helidon.common.socket.SocketContext log
WARNING: [0x6210162c 0x190bc5d8] Request failed: HttpPrologueRecord[protocol=HTTP, protocolVersion=1.1, method=GET, uriPath=/person/asd/asd, query=, fragment=], cannot send error response, as response already sent
io.helidon.common.http.RequestException: Unable to serialize property 'comments' from org.example.Person
	at io.helidon.common.http.RequestException$Builder.build(RequestException.java:126)
	at io.helidon.nima.webserver.http.ErrorHandlers.handleError(ErrorHandlers.java:117)
	at io.helidon.nima.webserver.http.ErrorHandlers.runWithErrorHandling(ErrorHandlers.java:76)
	at io.helidon.nima.webserver.http.Filters.filter(Filters.java:75)
	at io.helidon.nima.webserver.http.HttpRouting.route(HttpRouting.java:94)
	at io.helidon.nima.webserver.http1.Http1Connection.route(Http1Connection.java:254)
	at io.helidon.nima.webserver.http1.Http1Connection.handle(Http1Connection.java:140)
	at io.helidon.nima.webserver.ConnectionHandler.run(ConnectionHandler.java:120)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:577)
	at java.base/java.util.concurrent.ThreadPerTaskExecutor$ThreadBoundFuture.run(ThreadPerTaskExecutor.java:352)
	at java.base/java.lang.VirtualThread.run(VirtualThread.java:287)
	at java.base/java.lang.VirtualThread$VThreadContinuation.lambda$new$0(VirtualThread.java:174)
	at java.base/jdk.internal.vm.Continuation.enter0(Continuation.java:327)
	at java.base/jdk.internal.vm.Continuation.enter(Continuation.java:320)
Caused by: jakarta.json.bind.JsonbException: Unable to serialize property 'comments' from org.example.Person
	at org.eclipse.yasson.internal.serializer.ObjectSerializer.serializeInternal(ObjectSerializer.java:71)
	at org.eclipse.yasson.internal.serializer.AbstractContainerSerializer.serialize(AbstractContainerSerializer.java:107)
	at org.eclipse.yasson.internal.Marshaller.serializeRoot(Marshaller.java:147)
	at org.eclipse.yasson.internal.Marshaller.marshall(Marshaller.java:73)
	at org.eclipse.yasson.internal.Marshaller.marshall(Marshaller.java:101)
	at org.eclipse.yasson.internal.JsonBinding.toJson(JsonBinding.java:167)
	at io.helidon.nima.http.media.jsonb.JsonbWriter.write(JsonbWriter.java:86)
	at io.helidon.nima.http.media.jsonb.JsonbWriter.write(JsonbWriter.java:60)
	at io.helidon.nima.http.media.MediaContextImpl$CloseStreamWriter.write(MediaContextImpl.java:250)
	at io.helidon.nima.webserver.http.ServerResponseBase.send(ServerResponseBase.java:107)
	at org.example.HelloController$Route._person(HelloController$Route.java:43)
	at io.helidon.nima.webserver.http.HttpRouting$RoutingExecutor.doRoute(HttpRouting.java:343)
	at io.helidon.nima.webserver.http.HttpRouting$RoutingExecutor.execute(HttpRouting.java:305)
	at io.helidon.nima.webserver.http.ErrorHandlers.runWithErrorHandling(ErrorHandlers.java:50)
	... 11 more
Caused by: jakarta.json.bind.JsonbException: Error getting value on: org.example.Person@727ee88e
	at org.eclipse.yasson.internal.model.PropertyModel.getValue(PropertyModel.java:378)
	at org.eclipse.yasson.internal.serializer.ObjectSerializer.marshallProperty(ObjectSerializer.java:97)
	at org.eclipse.yasson.internal.serializer.ObjectSerializer.serializeInternal(ObjectSerializer.java:69)
	... 24 more
Caused by: java.lang.RuntimeException: foo
	at org.example.Person.getComments(Person.java:26)
	at org.eclipse.yasson.internal.model.PropertyModel.getValue(PropertyModel.java:376)
	... 26 more
```

-----
So to me it looks like there are a few options here:

A) Always write the json content to a separate buffer and then on success only write that to the server outputStream (this isn't great if the output is very large). This might be more a bug fix we would apply to avaje-jex rather than avaje-jsonb.
 
B) Treat as a bug in avaje-jsonb specifically in JsonType.toJson(bean, outputStream) ... remove the try-with-resources so that it ONLY flush() when successful + throw an exception indicating that it's written partial content to the OutputStream (maybe add JsonPartialException). If a JsonPartialException is thrown then this is an exception that an exception handler can't change the output stream content on. Note that currently there is the by default 4k buffer so it only writes to the outputStream when that 4k buffer needs to be written so for this test case because the content written before error is much smaller than 4k this would not end up writing anything to the underlying outputStream and the exception handler would work (but if the content prior to error was bigger than our 4k buffer size a JsonPartialException would be thrown and the jex error handler would not be able to over write the content (as some has been sent).

C) Give people the option to either always write to a separate buffer or write direct.

**Comment by SentryMan** [comment]
When I try this, it works as expected with a Javalin server with no exception handler.
```
  @Override
  public final void toJson(T value, OutputStream outputStream) {
    final JsonWriter writer = jsonb.writer(outputStream);
    adapter.toJson(writer, value);
    writer.close();
    close(outputStream);
  }
```
```
{
    "title": "Server Error",
    "status": 500,
    "type": "https://javalin.io/documentation#internalservererrorresponse",
    "details": {}
}```

**Comment by rbygrave** [comment]
Yes, when the `writer.close()` is only executed on success ... then the json content isn't flushed to the underlying outputStream.  (And this part we can fix in both avaje-jex and avaje-jsonb ... and we probably should fix it in avaje-jsonb).

**Comment by rbygrave** [comment]
To clarify a little, the default Jetty buffer is 32kb. We can write to that and as long as we don't call flush() or close() or exceed the 32kb buffer then Jetty will not send/commit the response (so it can set content-length etc).

The fix to avaje-jsonb is to not perform the underlying target flush() that it currently does. Also for the Jackson Core adapter we can get the similar result via disabling Jackson features on it's JsonGenerator:
```java
generator.disable(JsonGenerator.Feature.AUTO_CLOSE_TARGET);
generator.disable(JsonGenerator.Feature.FLUSH_PASSED_TO_STREAM);
generator.disable(JsonGenerator.Feature.AUTO_CLOSE_JSON_CONTENT); // optional really
``` 

The fix to avaje-jex is to: A) Use fixed avaje-jsonb + fix the JacksonJsonService. B) Fix the error handling to check for `isCommitted()` and `reset()` the underlying buffer such that the error handler can then write error content successfully.

So there are effectively 3 fixes to make this nice and work as we would like.

**Comment by rbygrave** [comment]
https://github.com/avaje/avaje-jsonb/pull/40

**Comment by rbygrave** [comment]
The second part of this fix is to avaje-jex via: https://github.com/avaje/avaje-jex/pull/44/files#diff-9ecf8be4fb89712f31eebb97168f222332f64ab411f7e90d3863e89110ed976bR31

**Comment by rbygrave** [comment]
I have released avaje-jex version 2.5-RC1 with this fix.

**Comment by rbygrave** [comment]
Ok, I'll close this issue now. We do need to release non-RC releases of avaje-jsonb and avaje-jex.
